crates/mun_hir/src/ty/infer/type_variable.rs

```diff
@@ -121,106 +190,112 @@ impl TypeVariableTable {
+    pub(crate) fn resolve_ty_completely(&mut self, ty: Ty) -> Ty {
+        self.resolve_ty_completely_inner(&mut Vec::new(), ty)
     }
 
-    /// Undoes all changes since the snapshot was created. Any snapshot created since that point
-    /// must already have been committed or rolled back.
-    pub fn rollback_to(&mut self, s: Snapshot) {
-        let Snapshot {
-            snapshot,
-            eq_snapshot,
-            mut bomb,
-        } = s;
-        self.values.rollback_to(snapshot);
-        self.eq_relations.rollback_to(eq_snapshot);
-        bomb.defuse();
-    }
-
-    /// Commits all changes since the snapshot was created, making them permanent (unless this
-    /// snapshot was created within another snapshot). Any snapshot created since that point
-    /// must already have been committed or rolled back.
-    pub fn commit(&mut self, s: Snapshot) {
-        let Snapshot {
-            snapshot,
-            eq_snapshot,
-            mut bomb,
-        } = s;
-        self.values.commit(snapshot);
-        self.eq_relations.commit(eq_snapshot);
-        bomb.defuse();
+    pub(crate) fn resolve_ty_completely_inner(
+        &mut self,
+        tv_stack: &mut Vec<TypeVarId>,
+        ty: Ty,
+    ) -> Ty {
+        ty.fold(&mut |ty| match ty {
+            Ty::Infer(tv) => {
+                let inner = tv.to_inner();
+                if tv_stack.contains(&inner) {
+                    return tv.fallback_value();
+                }
+                if let Some(known_ty) = self.eq_relations.inlined_probe_value(inner).known() {
+                    // known_ty may contain other variables that are known by now
+                    tv_stack.push(inner);
+                    let result = self.resolve_ty_completely_inner(tv_stack, known_ty.clone());
+                    tv_stack.pop();
+                    result
+                } else {
+                    tv.fallback_value()
+                }
+            }
+            _ => ty,
+        })
     }
-}
 
-impl SnapshotVecDelegate for Delegate {
-    type Value = TypeVariableData;
-    type Undo = Instantiate;
-
-    fn reverse(_values: &mut Vec<TypeVariableData>, _action: Instantiate) {
-        // We don't actually have to *do* anything to reverse an
-        // instantiation; the value for a variable is stored in the
-        // `eq_relations` and hence its rollback code will handle
-        // it. In fact, we could *almost* just remove the
-        // `SnapshotVec` entirely, except that we would have to
-        // reproduce *some* of its logic, since we want to know which
-        // type variables have been instantiated since the snapshot
-        // was started, so we can implement `types_escaping_snapshot`.
-        //
-        // (If we extended the `UnificationTable` to let us see which
-        // values have been unified and so forth, that might also
-        // suffice.)
-    }
+    // /// Returns indices of all variables that are not yet instantiated.
```

Comment:
Is this for future reference?

Reply:
Yes, I left it in there because it seems useful in the future.